### PR TITLE
🎉 Migrate Paypal Transaction Source Connector to Low Code

### LINKED ISSUE #29134: paypal transaction

### PR DESCRIPTION
**Pull request by topefolorunso**
## What
Migrating Source Paypal Transaction to Low-Code CDK resolves #29134

## How
Created the connector using the connector buider UI, then imported the manifest.yaml to the low code cdk.

## Recommended reading order
1. `manifest.yaml`

## 🚨 User Impact 🚨
*Are there any breaking changes? What is the end result perceived by the user?*

*For connector PRs, use this section to explain which type of semantic versioning bump occurs as a result of the changes. Refer to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines for more information. **Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-paypal-transaction/metadata.yaml [hunk 1]
```diff
@@ -3,26 +3,27 @@ data:
...
   githubIssueLabel: source-paypal-transaction
   icon: paypal.svg
   license: MIT
   name: Paypal Transaction
-  registries:
-    cloud:
-      enabled: true
-    oss:
-      enabled: true
+  releaseDate: TODO
```
Comment (sajarin) [inline airbyte-integrations/connectors/source-paypal-transaction/metadata.yaml:20]: Can you add a releaseDate?

File: airbyte-integrations/connectors/source-paypal-transaction/metadata.yaml [hunk 2]
```diff
@@ -3,26 +3,27 @@ data:
...
   releaseStage: generally_available
+  supportLevel: certified
   documentationUrl: https://docs.airbyte.com/integrations/sources/paypal-transaction
   tags:
-    - language:python
+    - language:lowcode
```
Comment (sajarin) [inline airbyte-integrations/connectors/source-paypal-transaction/metadata.yaml:25]: This should be low-code

**Review by sajarin — CHANGES_REQUESTED** [review]
Can you add a changelog entry?

**Comment by sajarin** [comment]
@topefolorunso there are some timestamp errors in the manifest.yaml. Please attach screenshots of passing acceptance tests

**Comment by topefolorunso** [comment]
> @topefolorunso there are some timestamp errors in the manifest.yaml. Please correct them and attach screenshots of passing acceptance tests

@sajarin errors should be fixed now. I'm unable to run integrations tests yet, waiting on @RealChrisSean for credentials

**Comment by sajarin** [comment]
image

Can you fix/update the sample config files in the integration_tests folder?

**Comment by topefolorunso** [comment]
> image
> Can you fix/update the sample config files in the integration_tests folder?

@sajarin Fixed now, please review

**Comment by sajarin** [comment]
Hey @topefolorunso thanks for the changes, now running into 401 errors with the connector, so need some time to investigate and debug.

**Comment by sajarin** [comment]
@topefolorunso according to the Paypal [docs](https://developer.paypal.com/api/rest/authentication/), the authentication request uses a -u parameter which I don't think is supported with our current HttpRequester component. 

If you take a look at the current version of the Paypal Transaction connector on master, there is a [custom implementation](https://github.com/airbytehq/airbyte/blob/d094b152c9d3fdde3f1d2cb1ea72c06cc94e1463/airbyte-integrations/connectors/source-paypal-transaction/source_paypal_transaction/source.py#L428C25-L428C25) of the Oauth2Authenticator class. I don't believe this is replicated with this migration. 

Take a look at the [RequestOptionsProvider](https://docs.airbyte.com/connector-development/config-based/understanding-the-yaml-file/request-options#request-options-provider), it might be possible to inject these values into the request payload

**Comment by topefolorunso** [comment]
@sajarin all fixed now. Kindly review.

There is no way to inject the credentials in auth header so I had to create a custom authenticator by extending the DeclarativeOauth2Authenticator class

**Comment by sajarin** [comment]
Running tests in a separate PR linked above

**Comment by topefolorunso** [comment]
@sajarin records are fetching correctly on my end. Maybe adjust the start_date config.

image

**Comment by topefolorunso** [comment]
> 2\. There seems to be a new config that's supported for the connector that is not supported in this migration. Can you add support for that new config type?

Can you point that out, please? I can't seem to locate it

**Comment by sajarin** [comment]
> @sajarin records are fetching correctly on my end. Maybe adjust the start_date config.
> 
> image

This is using our credentials or your own? I believe transactions is set to be an empty stream on our side. I'm having difficulty fetching balances records with this version of the connector. 


> > 2. There seems to be a new config that's supported for the connector that is not supported in this migration. Can you add support for that new config type?
> 
> Can you point that out, please? I can't seem to locate it

Take a look [here](https://github.com/airbytehq/airbyte/blob/72c4f5446f313d62a275051634eaae282a455c60/airbyte-integrations/connectors/source-paypal-transaction/source_paypal_transaction/source.py#L452)

**Comment by topefolorunso** [comment]
> This is using our credentials or your own?

Mine

> I'm having difficulty fetching balances records with this version of the connector.

How about I share my config.json via slack, can you use it?

**Comment by topefolorunso** [comment]
> > > 3\. There seems to be a new config that's supported for the connector that is not supported in this migration. Can you add support for that new config type?
> > 
> > 
> > Can you point that out, please? I can't seem to locate it
> 
> Take a look [here](https://github.com/airbytehq/airbyte/blob/72c4f5446f313d62a275051634eaae282a455c60/airbyte-integrations/connectors/source-paypal-transaction/source_paypal_transaction/source.py#L452)

Yes I noticed that but that logic is never implemented as there is no "credentials" property in the config. The logic is not consistent with the spec.json file [here](https://github.com/airbytehq/airbyte/blob/72c4f5446f313d62a275051634eaae282a455c60/airbyte-integrations/connectors/source-paypal-transaction/source_paypal_transaction/spec.json#L7-L42) and that's why I skipped it. The only possible auth method from the logic and spec is the oauth2

**Comment by sajarin** [comment]
> @sajarin
> 
> > This is using our credentials or your own?
> 
> Mine
> 
> > I'm having difficulty fetching balances records with this version of the connector.
> 
> How about I share my config.json via slack, can you use it?

We can't use your credentials in this case. We need to have this pass successfully with our credentials since that would confirm the migration was successful.

We do not have any transactions records in our account, which would explain why we're not able to fetch transaction records with the new low-code version. We do have balance records but we are not able to fetch balances records using the low-code for some reason. 

Given that the connector has some special conditions with the start_time, my guess is there is some missing functionality here with the request parameters that has not been migrated. It is difficult to tell what exactly though given the fact that the previous connector had logging statements and this one doesn't.

**Comment by topefolorunso** [comment]
@sajarin It's all fixed now. The issue was with the field path I defined for the balances stream. Please review.

**Comment by sajarin** [comment]
Thanks @topefolorunso for fixing the issue with the balances stream, fetching records properly now. There's an incremental test failure regarding future state, can you push a fix? 

image

Also one concern I had was about the transaction_intialization_date being used as the cursor for the transactions stream. Would transaction_updated_date be a better field for this? What do you think?

**Comment by topefolorunso** [comment]
> Thanks @topefolorunso for fixing the issue with the balances stream, fetching records properly now. There's an incremental test failure regarding future state, can you push a fix?
> 
> image

@sajarin Is this incremental test failure also for balances stream?

**Comment by topefolorunso** [comment]
> Also one concern I had was about the transaction_intialization_date being used as the cursor for the transactions stream. Would transaction_updated_date be a better field for this? What do you think?

Yes, you're correct. `transaction_updated_date` would be a better field to use

**Comment by topefolorunso** [comment]
Fixed now @sajarin

**Comment by marcosmarxm** [comment]
Hello, and thank you so much for contributing to the project during the Hackathon. Due to the high complexity of the code migration, we still have a few contributions open. I'm sending this message to update you that this contribution will be added to the final points we're going to announce tomorrow. With that said, we ask for your help to continue the process and finish the review.

**Review by sajarin — APPROVED** [review]

**Comment by topefolorunso** [comment]
> Thanks for contribution @topefolorunso!

Thank you @sajarin, my pleasure

**Pull request merged**